cds_ils/importer/eitems/importer.py

```diff
@@ -0,0 +1,169 @@
+from invenio_app_ils.errors import IlsValidationError
+from invenio_app_ils.proxies import current_app_ils
+from invenio_db import db
+
+from cds_ils.importer.config import CDS_ILS_IMPORTER_PROVIDERS
+from cds_ils.importer.eitems.api import get_eitems_for_document_by_provider
+
+
+class EItemImporter(object):
+    """EItem importer class."""
+
+    def __init__(
+        self,
+        json_metadata,
+        metadata_provider,
+        current_provider_priority,
+        delete_lower_priority_providers,
+        open_access,
+        login_required,
+    ):
+        """Constructor."""
+        self.json_data = json_metadata
+        self.metadata_provider = metadata_provider
+        self.current_provider_priority = current_provider_priority
+        self.delete_lower_priority_providers = delete_lower_priority_providers
+        self.open_access = open_access
+        self.login_required = login_required
+
+    def _get_record_import_provider(self, record):
+        """Get provider of the record."""
+        """Get an import provider of a given document."""
+        is_imported = record["created_by"]["type"] == "import"
+        if is_imported:
+            return record["created_by"]["value"]
+
+    def _set_record_import_source(self, record_dict):
+        """Set the provider of the record."""
+        record_dict["created_by"] = {
+            "type": "import",
+            "value": self.metadata_provider,
+        }
+
+    def _should_overwrite_eitem(self, eitem):
+        """Decide if the eitem should be replaced."""
+        existing_provider = self._get_record_import_provider(eitem)
+
+        if not existing_provider:
+            return False
+
+        existing_priority = CDS_ILS_IMPORTER_PROVIDERS[existing_provider][
+            "priority"
+        ]
+
+        # existing_priority = 0, self.priority = 1, returns False
+        return existing_priority >= self.current_provider_priority
+
+    def update_eitems(self, matched_document):
+        """Update eitems for given document."""
+        eitem_cls = current_app_ils.eitem_record_cls
+        eitem_indexer = current_app_ils.eitem_indexer
+        eitem_search = current_app_ils.eitem_search_cls()
+
+        document_pid = matched_document["pid"]
+
+        results = get_eitems_for_document_by_provider(
+            document_pid, self.metadata_provider
+        )
+        hits_total = results.hits.total.value
+        if hits_total == 0:
+            self.import_eitem(matched_document)
+        elif hits_total == 1:
+            try:
+                existing_eitem = eitem_cls.get_record_by_pid(
+                    results.hits[0].pid)
+                metadata_to_update = {}
+                self._build_eitem_dict(metadata_to_update,
+                                       matched_document["pid"])
+                existing_eitem.update(metadata_to_update)
+                existing_eitem.commit()
+                eitem_indexer.index(existing_eitem)
```

Comment:
indexing before `db.session.commit` might create the issue that if the db commit fails, it is already indexed. What about collecting all the records to index in a list and index at the end after the db commit?